File: hosts/ruby/lib/model_language.rb
# frozen_string_literal: true

# Ruby host for the model-language WebAssembly module. Runs the exact same engine
# as the TypeScript package: one JSON request in on the module's stdin, one JSON
# response out on its stdout. Same input, same output.

require "wasmtime"
require "json"
require "tempfile"

module ModelLanguage
  class Engine
    def initialize(wasm_path = nil)
      wasm_path ||= ENV["MODEL_LANGUAGE_WASM"] || bundled_wasm ||
                    File.expand_path("../../../wasm/dist/model_language.wasm", __dir__)
      @engine = Wasmtime::Engine.new
      @module = Wasmtime::Module.from_file(@engine, wasm_path)
      @linker = Wasmtime::Linker.new(@engine)
      Wasmtime::WASI::P1.add_to_linker_sync(@linker)
    end

    def render(template, data: {}, schema: [], options: {})
      invoke("op" => "render", "template" => template, "data" => data,
             "schema" => schema, "options" => options)
    end

    def validate(template, schema: [], options: {})
      invoke("op" => "validate", "template" => template, "schema" => schema, "options" => options)
    end

    def parse(template)
      invoke("op" => "parse", "template" => template)
    end

    private

    def invoke(request)
      Tempfile.create("model-language") do |out|
        config = Wasmtime::WasiConfig.new
                 .set_stdin_string(JSON.dump(request))
                 .set_stdout_file(out.path)
        store = Wasmtime::Store.new(@engine, wasi_p1_config: config)
        instance = @linker.instantiate(store, @module)

        error = nil
        begin
          instance.invoke("_start")
        rescue StandardError => e
          error = e # a clean WASI exit surfaces as an error; the response is written
        end

        content = File.read(out.path)
        raise error if content.empty? && error

        JSON.parse(content)
      end
    end

    def bundled_wasm
      path = File.expand_path("model_language.wasm", __dir__)
      File.exist?(path) ? path : nil
    end
  end
end

File: hosts/elixir/lib/model_language.ex
defmodule ModelLanguage do
  @moduledoc """
  Elixir bindings for model-language — a typed, safe template language for
  AI-agent prompts.

  Runs the exact same engine as the JavaScript/TypeScript package via an embedded
  WebAssembly module (`priv/model_language.wasm`), so every template renders
  byte-for-byte identically. Never crashes: template problems degrade to empty
  output plus a warning.

      iex> ModelLanguage.render("Hi {{ user.name | default: 'there' }}!", data: %{"user" => %{"name" => "Vasyl"}})["text"]
      "Hi Vasyl!"
  """

  @doc "Render `template` against `data`. Opts: `:data`, `:schema`, `:options`."
  def render(template, opts \\ []) do
    invoke(%{
      "op" => "render",
      "template" => template,
      "data" => Keyword.get(opts, :data, %{}),
      "schema" => Keyword.get(opts, :schema, []),
      "options" => Keyword.get(opts, :options, %{})
    })
  end

  @doc "Validate `template` against a schema. Opts: `:schema`, `:options`."
  def validate(template, opts \\ []) do
    invoke(%{
      "op" => "validate",
      "template" => template,
      "schema" => Keyword.get(opts, :schema, []),
      "options" => Keyword.get(opts, :options, %{})
    })
  end

  @doc "Parse `template` to its AST."
  def parse(template), do: invoke(%{"op" => "parse", "template" => template})

  defp wasm_bytes do
    path =
      System.get_env("MODEL_LANGUAGE_WASM") ||
        Path.join(Application.app_dir(:model_language, "priv"), "model_language.wasm")

    File.read!(path)
  end

  defp invoke(request) do
    {:ok, stdin} = Wasmex.Pipe.new()
    {:ok, stdout} = Wasmex.Pipe.new()
    Wasmex.Pipe.write(stdin, Jason.encode!(request))
    Wasmex.Pipe.seek(stdin, 0)

    wasi = %Wasmex.Wasi.WasiOptions{stdin: stdin, stdout: stdout}
    {:ok, pid} = Wasmex.start_link(%{bytes: wasm_bytes(), wasi: wasi})
    # A clean WASI exit returns an error; the response is already written.
    Wasmex.call_function(pid, "_start", [])

    Wasmex.Pipe.seek(stdout, 0)
    Jason.decode!(Wasmex.Pipe.read(stdout))
  end
end
